common/acl/camanager.go

```diff
@@ -0,0 +1,311 @@
+	cert, err := tls.LoadX509KeyPair(m.certFile, m.keyFile)
+	if err != nil {
+		return nil, err
+	}
+	if len(cert.Certificate) == 0 {
+		return nil, fmt.Errorf("cert.Certificate size is zero")
+	}
+	certs, err := x509.ParseCertificates(cert.Certificate[0])
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse certificate data: %w", err)
+	}
+	cert.Leaf = certs[0]
+	return &cert, nil
+}
+
+// applyCerts realizes the certificate application by token
+func (m *caManager) applyCerts() error {
+	cacert, err := GetCACert(m.caURL)
+	if err != nil {
+		return fmt.Errorf("failed to get CA certificate: %w", err)
+	}
+
+	// validate the CA certificate by hashcode
+	// token's format: <sha256.Sum256(cacert)>.<jwt token>, and jwt token has three parts split by '.'
+	tokenParts := strings.Split(m.token, ".")
+	if len(tokenParts) != 4 {
+		return fmt.Errorf("token credentials are in the wrong format, token: %s", m.token)
+	}
+	hash, newHash, ok := ValidateCACerts(cacert, tokenParts[0])
+	if !ok {
+		return fmt.Errorf("failed to validate CA certificate. tokenCAhash: %s, CAhash: %s", hash, newHash)
+	}
+
+	// save the ca.crt to file
+	ca, err := x509.ParseCertificate(cacert)
+	if err != nil {
+		return fmt.Errorf("failed to parse the CA certificate, error: %w", err)
+	}
+
+	if err = certutil.WriteCert(m.caFile, ca); err != nil {
+		return fmt.Errorf("failed to save the CA certificate to file: %s, error: %v", m.caFile, err)
+	}
+
+	// get the edge.crt
+	caPem := pem.EncodeToMemory(&pem.Block{Bytes: cacert, Type: cert.CertificateBlockType})
+	pk, edgeCert, err := m.GetNodeCert(m.certURL, caPem, tls.Certificate{}, strings.Join(tokenParts[1:], "."))
+	if err != nil {
+		return fmt.Errorf("failed to get edge certificate from the cloudcore, error: %v", err)
+	}
+
+	// save the edge.crt to the file
+	crt, _ := x509.ParseCertificate(edgeCert)
+	if err = certutil.WriteKeyAndCert(m.keyFile, m.certFile, pk, crt); err != nil {
+		return fmt.Errorf("failed to save the edge key and certificate to file: %s, error: %v", m.certFile, err)
+	}
+
+	return nil
+}
+
+// generateKey realizes the acl application by token
+func (m *caManager) generateKey() error {
+	pk, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return fmt.Errorf("failed to generate key, error: %v", err)
+	}
+	// save the private key to the file
+	if err = certutil.WriteKey(m.keyFile, pk); err != nil {
+		return fmt.Errorf("failed to save the private key %s, error: %v", m.keyFile, err)
+	}
+	return nil
+}
+
+// getCA returns the CA in pem format.
+func (m *caManager) getCA() ([]byte, error) {
+	return ioutil.ReadFile(m.caFile)
+}
+
+// GetNodeCert applies for the certificate from cloudcore
+func (m *caManager) GetNodeCert(url string, capem []byte, cert tls.Certificate, token string) (*ecdsa.PrivateKey, []byte, error) {
+	pk, csr, err := m.getCSR()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create CSR: %v", err)
```

Comment:
inconsistent: since you call `m.getCSR` instead of createCSR
```suggestion
		return nil, nil, fmt.Errorf("failed to get CSR: %v", err)
```